Governance updates (#3490)
withastro/astro: GOVERNANCE.md

# Governance

Herein contains Astro's goverance model. It outlines and describes in detail each role and their subsequent responsibilties. This includes; 

- Nomination Process
- Duty of Care
- Code Review Process
- Code of Conduct Enforcement
  
This document outlines the governance model for Astro. This includes detailed descriptions of different roles, nomination processes, code review processes, and Code of Conduct enforcement.

> 🚨 **All community members must follow the [Code of Conduct (CoC)](CODE_OF_CONDUCT.md).**  
Consequences for CoC violations are detailed in [Moderation](#moderation).

> 👉 **Want to trigger a vote, nomination, or perform some other action?**  
Scroll down to [Playbook](#governance-playbook).

## Get Involved

**Any contribution that helps support or further the Astro project and its wider community is always welcome and greatly appreciated.**
 

All types of contribution are meaningful, from code to documentation, even blog posts. 

**Anyone** can become an Astro Contributor (yes, even you!). Regardless of skill, experience or background.

> Our goal is too always recognize all of our contributors and their effort given to Astro. 

Reviewer: ```suggestion
Our goal is to always recognize all of our contributors and their effort given to Astro. 
```

Reviewer 2: 🤦‍♂️ D'Oh

> 🚀 Remember no contribution is too small or too little. 

We will always recognise and value all our contributors and their effort that they bring to the Astro project. 

Most importantly we are eternally grateful for the vast and diverse impact that they bring to our community.

## Contributor Levels

We recognize different degrees of contributions made to the Astro project as **Contributor Levels** also referred to as **Contributor Roles**. 

Contributor levels are available to **all members** within the Astro Community, regardless of coding skill or experience. 

The two off the most important things that we look from our contributors is:

- **Being here** - Everyone's time is valuable, and the fact that you're here and contributing to Astro is amazing! Thank you for being a part of this project with us.
  
- **Being a positive member of our community**  
  Strive to go above and beyond our [Code of Conduct](CODE_OF_CONDUCT.md) and commit to a healthy, honest discourse when communicating: 
  - Within our Github Project and its wider umbrella groups and projects.
    - Pull Requests, 

>     - Issue's 

Reviewer: ```suggestion
    - Issues 
```

    - Discussions,
    - RFC process, 
  - Discord conversations, 
  - Any public and outward interactions conducted outside of our community (ex: no Twitter bullies allowed 🤼‍♂️)

All Contributor roles are granted for as long as the individual wishes to engage with the project. 

> Contributor's can voluntarily leave the project at any time (see [Retiring a Role](#retiring-a-role-alumni) below). 

Reviewer: ```suggestion
Contributors can voluntarily leave the project at any time (see [Retiring a Role](#retiring-a-role-alumni) below). 
```

#### Responsibilities

This role does not require any extra responsibilities nor time commitment. We hope you stick around and keep participating in our community!

If you're interested in reaching the next level and becoming a **Maintainer**, you can explore some of those responsibilities in the [next section](#level-2-l2---maintainer).

#### Nomination Process

> You may self-nominate by sending the message `!contribute` in any Discord channel. If you do this, please share a second message with a link or description of your contribution so that people can recognize you for the contribution,

Reviewer 3: ```suggestion
You may self-nominate by sending the message `!contribute` in any Discord channel. If you do this, please share a second message with a link or description of your contribution so that people can recognize you for the contribution.
```

- All privileges of the [Contributor role](#level-1---contributor), plus...
- Invitation to the `@maintainer` role on [Discord](https://astro.build/chat)
- Invitation to the private `#maintainers` channel on Discord.
- Invitation to the `@maintainers` team on GitHub.
- New name color on Discord: **blue**.
- Ability to moderate Discord to remove spam, harmful speech, etc.
- Ability to join the `@mods` role on Discord (optional, opt-in).

> - Ability to push branches directly to the `astro` GitHub repo (No more personal fork needed).

Reviewer 2: I HAve added those changes in, `astro` isnt a valid repo and this better reflects the Organisation on Github

> - Ability to push branches directly to the `astro` GitHub repo (No more personal fork needed).

Author: yup, good call out

#### Nomination

- To be nominated, a nominee is expected to already be performing some of the responsibilities of a Maintainer.
- You can be nominated by any existing Maintainer (L2 or above). 
- Once nominated, there will be a vote by existing Maintainers. 
- See [vote rules & requirements](#voting) for info on how the vote works.

### Level 3 (L3) - Core

> The **Core** role is available to community members who have a larger-than-usual impact on the Astro project and community. They are seen as leaders in the project and are listened to by the wider Astro community, often before they have even reached this level. A Core member is recognized for contributing a significant amount of time and energy to the project through issues, pull requests, bug fixes, implementing advanced enhancements/features, and/oractivly posting on Discord. 

Reviewer 2: ```suggestion
The **Core** role is available to community members who have a larger-than-usual impact on the Astro project and community. They are seen as leaders in the project and are listened to by the wider Astro community, often before they have even reached this level. A Core member is recognized for contributing a significant amount of time and energy to the project through issues, pull requests, bug fixes, implementing advanced enhancements/features, and/or activly posting on Discord. 
```

Reviewer 2: Just a wee nit,

Author: thanks!

#### Nomination

- To be nominated, a nominee is expected to already be performing some of the duties and responsibilities of a Maintainer.
- You can only be nominated by any existing Maintainer (L2 or above). 
- Once nominated, there will be a vote by existing Maintainers. 
- See [vote rules & requirements](#voting) for info on how the vote works.

### Level 3 (L3) - Core

The **Core** role is available to community members who have a larger-than-usual impact on the Astro project and community. 

They are seen as leaders in the project and are listened to by the wider Astro community, often before they have even reached this level. 

A Core member is recognized for contributing a significant amount of their time and energy to the project through issues, pull requests, bug fixes, implementing advanced enhancements/features, and is activly posting on Discord. 

> Not every contributor will reach this level, and that's okay! L2 Maintainers still have significant responsibility and privileges as is within our community.

Reviewer: ```suggestion
Not every contributor will reach this level, and that's okay! L2 Maintainers still have significant responsibility and privileges within our community.
```

Reviewer 2: As in where did this come from.... like I have no excuses for this utter brain fart💨

#### Nomination

- Stewards cannot be self-nominated.
- Only Core members are eligible.
- New Stewards will be added based on a unanimous vote by the existing Steward(s).
- In the event that someone is unreachable then the decision will be deferred.

## Other Roles

### Project Teams

Besides our contributor levels described above, there are additional roles and teams available that community members are welcome to join. Roles are a great way to organize around different projects and initiatives in our community. For example:

- `@team-docs` runs the `#docs` channel and organizes the growth and development of Astro documentation.
- `@i18n-gang` runs the `#docs-i18n` channel and organizes translations in several languages.
- `@support-squad` runs the `#support-threads` channel and helps anyone who needs help using Astro.

> Many of these team roles can be be browsed and joined automatically by visiting the `#manage-roles` channel in our Discord. Getting involved with a team is a great way to start contributing to Astro!

Reviewer 4: A small typo I found:
```suggestion
Many of these team roles can be browsed and joined automatically by visiting the `#manage-roles` channel in our Discord. Getting involved with a team is a great way to start contributing to Astro!
```

Author: thank you @Yan-Thomas!

### Moderator

**Moderator** is a special role available to Maintainers (L2 and above). While all maintainers are granted permissions to moderate for bad behavior across our community, a Moderator actively takes on this the responsibility. For example, a community member may ping moderators (via the `@mods` role) to resolve spam posts or Code of Conduct violations.

Trivial tasks (like removing spam) can be acted on unilaterally by a Moderator. Other non-trivial tasks (like assisting with or resolving a Code of Conduct violation) should involve the entire Moderator team (and in some cases, the project Steward).

#### Privileges

- `@mods` role on [Discord](https://astro.build/chat)
- Invitation to the private `#moderators` channel on Discord
- Invitation to the `staff` team on GitHub.

#### Nomination

Any Maintainer (L2 and above) can self-nominate by messaging the project Steward (`@steward`) on Discord.

> ### Technical Steering Committee (TSC)

Reviewer 5: So does "member of the technical steering committee" end up being like a "super core member" in a sense, requiring expertise in the code base?

I'm sorry, I hadn't seen the reason for this change

Reviewer 6: The reason is so that people who are heavily involved but not working on the codebase can still become part of core. So the TSC becomes what the current core sort of was; for people working on the Astro codebase and who can guide the technical direction of the project.

Author: yup, less "super core member" and more "one team of many". It is the only team right now with extra voting permissions in our goverance docs, but this aligns `tsc` closer to a project like `docs` or `support-squad`. 

It also opens the door for other teams to adopt their own special governance down the road, similar to TSC. For example, if there ever was a reason to add a voting mechanism for the docs team, we could define the docs voting system in this same place.

Reviewer 5: hmm ok so trying to go in the direction of working groups, kinda?

tsc = team focused on the codebase
docs = team focused on the docs
support-squad = team focused on providing support in discord & github & more

#### Responsibilities 

- Participating in RFC discussions and technical meetings.
- Assisting with design and implementation of non-trivial GitHub PRs.
- Reviewing and merging larger, non-trivial PRs.
- Maintaining and improving overall codebase architecture.
- Tracking and ensuring progress of open pull requests.
- Mentoring and guiding other community contributors.

#### Nomination

- To be nominated, a nominee is expected to already be active in technical discussions and performing some of the responsibilities of a TSC member.
- You can be nominated by any existing Core member (L3 or above). Note: This includes all existing TSC members as well.
- Once nominated, there will be a vote by existing Core members.
- See [vote rules & requirements](#voting) for info on how this vote works.

### Staff

**Staff** is a special designation for employees of [The Astro Technology Company](https://astro.build/company). 

#### Privileges

- `@staff` role on [Discord](https://astro.build/chat)
- New name color on Discord: **yellow**.
- Invitation to some private channels on Discord, at the discretion of the project Steward.
- Invitation to the `staff` team on GitHub.

> Staff membership does not grant any additional abilities when it comes to voting and project governance. A Staff member is still eligible for other roles in the community and may still vote as defined by their other roles. For example, a Staff member who is also a part of `@core` will be able to vote as any other `@core` member would.

Author: Not really! This flow you're describing is mainly covered by the "Core In Residence" role. 
For 99% of people, I imagine it will work like this:

- Someone joins The Astro Technology Company to work on the core codebase full-time
- They are nominated for "Core In Residence", meaning that they get the `@core` status but can not participate in voting.
- They leave the company and lose their core status. If they want to stick around our community, they can go through the normal nomination channels.

If someone joins the company but does not work in the community, they may get staff but wouldn't even be nominationed for maintainer, core, etc.

If someone joins the company and contributes to the codebase for many years, becoming a leader in the community, then they could have their "in residence" limitations removed by a normal core nomination vote.

Hope that helps clarify!

> Staff membership does not grant any additional abilities when it comes to voting and project governance. A Staff member is still eligible for other roles in the community and may still vote as defined by their other roles. For example, a Staff member who is also a part of `@core` will be able to vote as any other `@core` member would.

Reviewer 5: Ah, so basically this is kinda switching the Staff role to a relatively decorative role + "core in residence" is an option to fast-track permissions for someone in the company but who hasn't necessarily been around long enough to earn any of the other statuses?

> Staff membership does not grant any additional abilities when it comes to voting and project governance. A Staff member is still eligible for other roles in the community and may still vote as defined by their other roles. For example, a Staff member who is also a part of `@core` will be able to vote as any other `@core` member would.

Author: yup, exactly.

### Alumni

**Alumni** is a special designation for Maintainers (L2 and above) who have stepped away from the project and no longer contribute regularly. See [Retiring a Role](#retiring-a-role-alumni) below for more information.

#### Privileges

- `@alumni` role on [Discord](https://astro.build/chat)
- New name color on Discord: **light blue**.
- Invitation to the private `#alumni` channel on Discord.

## Retiring a Role (Alumni)

Contributor roles are granted for as long as the person wishes to engage with the project. However, over time an active community member may choose to step away from the Astro project to work on other things. Moving on from a project is a natural and well-understood part of any open source community, and we celebrate it!

> **Alumni** is a special designation and role for any person who was once an active maintainer (L2 or above) but is now no longer actively involved. By retiring and joining Alumni you trade-in your current set of roles, privileges, and responsibilities for a new, special Alumni role (which comes with its own set of Privileges, as described above).

Reviewer 5: What is the role / privilege of being an Alumni? I imagine it's a way to communicate that you have participated in helping before, but that you're taking a break for some amount of time and aren't involved enough?

Author: That's correct! No additional privileges other than what is outlined here.

2. A trivial change can be discussed and approved entirely within the RFC GitHub issue, as long as there are no objections from Core or TSC members. This is not considered a formal vote.
3. A non-trivial, significant change should be discussed within the RFC and approved during an RFC meeting call. In some cases, an RFC may be approved outside of an RFC meeting using Pull Request reviews as a proxy for votes.
4. During an RFC meeting, the person leading the call will attempt to achieve consensus on the RFC proposal.
5. **If consensus is reached:** the RFC is approved.
6. **If consensus is not reached:** The RFC author and TSC members must make all reasonable attempts to resolve issues and reach consensus in GitHub or a follow-up RFC meeting. The process of reaching consensus can take time, and should not be rushed as long as all participants are making a reasonable effort to respond.
7. **If consensus still cannot be reached:** The project Steward may invoke [rough consensus](https://en.wikipedia.org/wiki/Rough_consensus) to resolve an RFC that has not achieved absolute consensus, as described below (borrowed from the [IETF](https://datatracker.ietf.org/doc/html/rfc2418)):

> > Working groups make decisions through a "rough consensus" process. Astro consensus does not require that all participants agree although this is, of course, preferred. In general, the dominant view of the TSC shall prevail. (However, "dominance" is not to be determined on the basis of volume or persistence, but rather a more general sense of agreement). Consensus can be determined by a show of hands, humming, or any other means on which the TSC agrees (by rough consensus, of course). Note that 51% of the TSC does not qualify as "rough consensus" and 99% is better than rough. It is up to the project Steward to determine if rough consensus has been reached.

Reviewer 5: Who determines rough consensus if an RFC is discussed on an RFC meeting w/o the Steward present?

I suppose that just comes down to what sort of RFC it is, how fundamental it is?

Author: It's up the the steward to determine if rough consensus is reached, so they would have to be present in any case of non-absolute consensus.

In practice, all of our votes so far have been through absolute consensus (100%), and any objections have been addressable through follow-up and working together between the RFC champion and the objector. So this is a bit of an edge case, but still a good one to plan for.

Reviewer 5: This extends to the "let's wait a moment for someone to object to consensus" done in RFC calls traditionally, right?

Author: yup!
